src/main/java/com/ecore/roles/service/impl/MembershipsServiceImpl.java

```diff
@@ -8,48 +9,62 @@
 import org.springframework.stereotype.Service;
 
+import javax.validation.Valid;
 import java.util.List;
+import java.util.Objects;
 import java.util.UUID;
 
-import static java.util.Optional.ofNullable;
-
 @Log4j2
 @Service
+@RequiredArgsConstructor
 public class MembershipsServiceImpl implements MembershipsService {
 
     private final MembershipRepository membershipRepository;
     private final RoleRepository roleRepository;
-
-    @Autowired
-    public MembershipsServiceImpl(
-            MembershipRepository membershipRepository,
-            RoleRepository roleRepository) {
-        this.membershipRepository = membershipRepository;
-        this.roleRepository = roleRepository;
-    }
+    private final TeamsService teamsService;
 
     @Override
-    public Membership assignRoleToMembership(@NonNull Membership m) {
-
-        UUID roleId = ofNullable(m.getRole()).map(Role::getId)
-                .orElseThrow(() -> new InvalidArgumentException(Role.class));
+    public Membership assignRoleToMembership(@NonNull @Valid Membership membership) {
+        validateResourceExists(membership);
+        validateTeamMember(membership);
+        return membershipRepository.save(membership);
+    }
 
-        if (membershipRepository.findByUserIdAndTeamId(m.getUserId(), m.getTeamId())
-                .isPresent()) {
+    private void validateResourceExists(Membership membership) {
+        UUID roleId = membership.getRole().getId();
+        Role role = roleRepository.findById(roleId)
+                .orElseThrow(() -> new ResourceNotFoundException(Role.class, roleId));
+        if (membershipRepository.existsByUserIdAndTeamIdAndRole(membership.getUserId(),
+                membership.getTeamId(), role)) {
             throw new ResourceExistsException(Membership.class);
         }
+    }
 
-        roleRepository.findById(roleId).orElseThrow(() -> new ResourceNotFoundException(Role.class, roleId));
-        return membershipRepository.save(m);
+    private void validateTeamMember(Membership membership) {
+        Team team = teamsService.getTeam(membership.getTeamId());
+        if (team == null) {
+            throw new ResourceNotFoundException(Team.class, membership.getTeamId());
+        }
+        Objects.requireNonNull(team).getTeamMemberIds().stream()
+                .filter(p -> p.equals(membership.getUserId()))
+                .findAny()
+                .orElseThrow(() -> new InvalidArgumentException(
+                        "Invalid 'Membership' object. The provided user doesn't belong to the provided team."));
+    }
+
+    @Override
+    public List<Membership> getMemberships(@NonNull UUID roleId) {
+        return membershipRepository.findByRoleId(roleId);
     }
 
     @Override
-    public List<Membership> getMemberships(@NonNull UUID rid) {
-        return membershipRepository.findByRoleId(rid);
+    public List<Role> getRoles(@NonNull UUID userId, @NonNull UUID teamId) {
```

Comment:
This method has been created to return the list of roles by userId and teamId raised by business rule.